npm/qsharp/ux/circuit.tsx

```diff
@@ -12,56 +13,157 @@ import { CircuitProps } from "./data.js";
 const MAX_OPERATIONS = 10000;
 const MAX_QUBITS = 1000;
 
-/* This component is shared by the Python widget and the VS Code panel */
+// This component is shared by the Python widget and the VS Code panel
 export function Circuit(props: { circuit: qviz.Circuit }) {
+  const circuit = props.circuit;
+  const unrenderable =
+    circuit.qubits.length === 0 ||
+    circuit.operations.length > MAX_OPERATIONS ||
+    circuit.qubits.length > MAX_QUBITS;
+
+  return (
+    <div>
+      {unrenderable ? (
+        <Unrenderable
+          qubits={props.circuit.qubits.length}
+          operations={props.circuit.operations.length}
+        />
+      ) : (
+        <ZoomableCircuit circuit={props.circuit} />
+      )}
+    </div>
+  );
+}
+
+function ZoomableCircuit(props: { circuit: qviz.Circuit }) {
   const circuitDiv = useRef<HTMLDivElement>(null);
+  const [zoomLevel, setZoomLevel] = useState<number | undefined>(undefined);
+  const [rendering, setRendering] = useState(true);
+
+  useEffect(() => {
+    // Enable "rendering" text while the circuit is being drawn
+    setRendering(true);
+    const container = circuitDiv.current!;
+    container.innerHTML = "";
+  }, [props.circuit]);
+
+  useEffect(() => {
+    if (rendering) {
+      const container = circuitDiv.current!;
+      // Draw the circuit - may take a while for large circuits
+      const svg = renderCircuit(props.circuit, container);
+      // Calculate the initial zoom level based on the container width
+      const initialZoom = calculateZoomToFit(container, svg as SVGElement);
+      // Set the initial zoom level
+      setZoomLevel(initialZoom);
+      // Resize the SVG to fit
+      updateWidth();
+      // Disable "rendering" text
+      setRendering(false);
+    }
+  }, [rendering]);
+
+  useEffect(() => {
+    updateWidth();
+  }, [zoomLevel]);
 
+  return (
+    <div>
+      <div>
+        <ZoomControl zoom={zoomLevel} onChange={setZoomLevel} />
+      </div>
+      <div>
+        {rendering
+          ? `Rendering diagram with ${props.circuit.operations.length} gates...`
+          : ""}
+      </div>
+      <div class="qs-circuit" ref={circuitDiv}></div>
+    </div>
+  );
+
+  function updateWidth() {
+    const svg = circuitDiv.current?.getElementsByClassName("qviz")[0];
+    if (svg) {
+      const width = svg.getAttribute("width")!;
+      svg.setAttribute(
+        "style",
+        `max-width: ${width}; width: ${(parseInt(width) * (zoomLevel || 100)) / 100}; height: auto`,
```

Comment:
This is interesting/confusing. Initially I thought this might be recursively modifying the value, as you are getting the width on line 87, then setting it here, so on second run it would scale the prior set value, not the original value.

But on closer reading I guess the `svg` element itself has a `width` attribute, but you are then setting `width` via the CSS `style` attribute here - leaving the original `width` attribute untouched.

Is that correct (and deliberate)? Worth some comments if so, as it's subtle.

Reply:
Yes, that's exactly it. I think I already have a comment about this somewhere but I'll add a better explanation here.